ddtrace/debugging/_function/store.py

```diff
@@ -0,0 +1,107 @@
+from types import CodeType
+from types import FunctionType
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import cast
+
+from ddtrace.debugging._function.discovery import FullyNamed
+from ddtrace.internal.injection import HookInfoType
+from ddtrace.internal.injection import HookType
+from ddtrace.internal.injection import eject_hooks
+from ddtrace.internal.injection import inject_hooks
+from ddtrace.internal.wrapping import Wrapper
+from ddtrace.internal.wrapping import WrapperFunction
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+WrapperType = Callable[[FunctionType, Any, Any, Any], Any]
+
+
+class FullyNamedWrapperFunction(FullyNamed, WrapperFunction):
+    """A fully named wrapper function."""
+
+
+class FunctionStore(object):
+    """Function object store.
+
+    This class provides a storage layer for patching operations, allowing to
+    store the original code object of functions being patched with either hook
+    injections or wrapping.
+
+    If extra attributes are defined during the patching process, they will get
+    removed when the functions are restored.
+    """
+
+    def __init__(self, extra_attrs=None):
+        # type: (Optional[List[str]]) -> None
+        self._code_map = {}  # type: Dict[FunctionType, CodeType]
+        self._extra_attrs = ["__dd_wrapped__"]
+        if extra_attrs:
+            self._extra_attrs.extend(extra_attrs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        self.restore_all()
+
+    def _store(self, function):
+        # type: (FunctionType) -> None
+        if function not in self._code_map:
+            self._code_map[function] = function.__code__
+
+    def inject_hooks(self, function, hooks):
+        # type: (FullyNamedWrapperFunction, List[Tuple[HookType, int, Any]]) -> None
+        """Bulk-inject hooks into a function."""
+        try:
+            self.inject_hooks(cast(FullyNamedWrapperFunction, function.__dd_wrapped__), hooks)
```

Comment:
So it keeps casting the wrapped functions until it reaches the non-wrapped version of the function, and then actually injects to hooks.